server-rlimit-nix.go

```diff
@@ -43,3 +43,34 @@ func setMaxOpenFiles() error {
 	}
 	return nil
 }
+
+// Set max memory used by minio as a process, this value is usually
+// set to 'unlimited' but we need to validate additionally to verify
+// if any hard limit is set by the user, in such a scenario would need
+// to reset the global max cache size to be 80% of the hardlimit set
+// by the user. This is done to honor the system limits and not crash.
+func setMaxMemory() error {
+	var rLimit syscall.Rlimit
+	err := syscall.Getrlimit(syscall.RLIMIT_AS, &rLimit)
+	if err != nil {
+		return err
+	}
+	// Set the current limit to Max, it is default 'unlimited'.
+	// TO decrease this limit further user has to manually edit
+	// `/etc/security/limits.conf`
+	rLimit.Cur = rLimit.Max
+	err = syscall.Setrlimit(syscall.RLIMIT_AS, &rLimit)
+	if err != nil {
+		return err
+	}
+	err = syscall.Getrlimit(syscall.RLIMIT_AS, &rLimit)
+	if err != nil {
+		return err
+	}
+	// Validate if rlimit memory is set to lower
+	// than max cache size. Then we should use such value.
+	if rLimit.Cur < globalMaxCacheSize {
```

Comment:
if rLimit.Cur is `RLIM_INFINITY` then we should limit globalMaxCacheSize so that it is < RAM size